source_gen/lib/src/generator.dart

```diff
@@ -18,11 +18,15 @@ abstract class Generator {
 
   /// Generates Dart code for an input Dart library.
   ///
-  /// May create additional outputs through the `buildStep`, but the 'primary'
-  /// output is Dart code returned through the Future. If there is nothing to
-  /// generate for this library may return null, or a Future that resolves to
-  /// null or the empty string.
-  FutureOr<String> generate(LibraryReader library, BuildStep buildStep) => null;
+  /// Supported return values include a single [String] or multiple [String]
```

Comment:
I'm not sure the reliance on Doc comments over type signatures is worth it in this case. How does this relate to the goal of a CombiningBuilder? I don't think that implementation needs to use the Generator infrastructure at all.

Reply:
It turns out to be *super* useful in the single case.

For instance, you can now emit helpers for each element in a generator and they get collapsed into one. So you don't need to (1) put them inline in a function or (2) put them in the "annotations" package.

I have a design of the combining builder as just a natural extension of the existing Part/Library builder impl, but I wanted to get this in first.

Reply:
@grouma was going to take a look at the CombiningBuilder

>For instance, you can now emit helpers for each element in a generator and they get collapsed into one. So you don't need to (1) put them inline in a function or (2) put them in the "annotations" package.

I'm not entirely convinced that emitting helpers this way is the most understandable way to handle it. It might work smoothly for us when we have a deep understanding of this package but  expands the surface area in a way that might be hard to document for other users.

> I have a design of the combining builder as just a natural extension of the existing Part/Library builder impl, but I wanted to get this in first.

The CombiningBuilder shouldn't need a `LibraryReader` (though the overall builder will probably want to check for the `part` directive and will want to use a Resolver) and I don't quite understand how it fits the `Generator` interface.

Reply:
Let's maybe chat about this when you're back in the office, or @grouma could review this if he thinks it will be useful.

Reply:
I don't have the context yet to determine if this will be useful. Let's hold off on submitting it. I'll be diving into https://github.com/dart-lang/source_gen/issues/319 today.